In [6]:
from keras.datasets import imdb
from keras.utils import pad_sequences

In [7]:
max_features = 5000  
max_len = 500  
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)

In [8]:
x_train = pad_sequences(x_train, maxlen=max_len)
x_test = pad_sequences(x_test, maxlen=max_len)

In [9]:
from keras.models import Sequential
from keras.layers import Embedding, SimpleRNN, Dense

In [11]:
model_rnn = Sequential()
model_rnn.add(Embedding(max_features, 32))  
model_rnn.add(SimpleRNN(32))  
model_rnn.add(Dense(1, activation='sigmoid'))  

In [12]:
model_rnn.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['accuracy'])
model_rnn.fit(x_train, y_train, epochs=10, batch_size=128, validation_split=0.2)

Epoch 1/10
157/157 [==============================] - 15s 84ms/step - loss: 0.6266 - accuracy: 0.6302 - val_loss: 0.5292 - val_accuracy: 0.7540
Epoch 2/10
157/157 [==============================] - 13s 80ms/step - loss: 0.4096 - accuracy: 0.8253 - val_loss: 0.4072 - val_accuracy: 0.8228
Epoch 3/10
157/157 [==============================] - 13s 83ms/step - loss: 0.3213 - accuracy: 0.8673 - val_loss: 0.3529 - val_accuracy: 0.8542
Epoch 4/10
157/157 [==============================] - 13s 80ms/step - loss: 0.2705 - accuracy: 0.8921 - val_loss: 0.4088 - val_accuracy: 0.8134
Epoch 5/10
157/157 [==============================] - 13s 82ms/step - loss: 0.2597 - accuracy: 0.9005 - val_loss: 0.4202 - val_accuracy: 0.8128
Epoch 6/10
157/157 [==============================] - 13s 85ms/step - loss: 0.2068 - accuracy: 0.9215 - val_loss: 0.5757 - val_accuracy: 0.7744
Epoch 7/10
157/157 [==============================] - 14s 87ms/step - loss: 0.1590 - accuracy: 0.9430 - val_loss: 0.5310 - val_accuracy:

In [13]:
from keras.layers import LSTM

model_lstm = Sequential()
model_lstm.add(Embedding(max_features, 32))
model_lstm.add(LSTM(32))
model_lstm.add(Dense(1, activation='sigmoid'))

In [14]:
model_lstm.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['accuracy'])
model_lstm.fit(x_train, y_train, epochs=10, batch_size=128, validation_split=0.2)

Epoch 1/10
157/157 [==============================] - 26s 150ms/step - loss: 0.5082 - accuracy: 0.7570 - val_loss: 0.3731 - val_accuracy: 0.8496
Epoch 2/10
157/157 [==============================] - 24s 152ms/step - loss: 0.3148 - accuracy: 0.8733 - val_loss: 0.3376 - val_accuracy: 0.8552
Epoch 3/10
157/157 [==============================] - 23s 148ms/step - loss: 0.2668 - accuracy: 0.8965 - val_loss: 0.3032 - val_accuracy: 0.8734
Epoch 4/10
157/157 [==============================] - 24s 150ms/step - loss: 0.2352 - accuracy: 0.9105 - val_loss: 0.2910 - val_accuracy: 0.8810
Epoch 5/10
157/157 [==============================] - 23s 150ms/step - loss: 0.2161 - accuracy: 0.9187 - val_loss: 0.4137 - val_accuracy: 0.8508
Epoch 6/10
157/157 [==============================] - 24s 153ms/step - loss: 0.2023 - accuracy: 0.9229 - val_loss: 0.3251 - val_accuracy: 0.8684
Epoch 7/10
157/157 [==============================] - 24s 156ms/step - loss: 0.1906 - accuracy: 0.9298 - val_loss: 0.3199 - val_ac

In [15]:
from keras.layers import Bidirectional

model_bidirectional = Sequential()
model_bidirectional.add(Embedding(max_features, 32))
model_bidirectional.add(Bidirectional(LSTM(32)))  
model_bidirectional.add(Dense(1, activation='sigmoid'))

In [16]:
model_bidirectional.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['accuracy'])
model_bidirectional.fit(x_train, y_train, epochs=10, batch_size=128, validation_split=0.2)

Epoch 1/10
157/157 [==============================] - 43s 248ms/step - loss: 0.5334 - accuracy: 0.7430 - val_loss: 0.3648 - val_accuracy: 0.8520
Epoch 2/10
157/157 [==============================] - 38s 242ms/step - loss: 0.3277 - accuracy: 0.8695 - val_loss: 0.4099 - val_accuracy: 0.8392
Epoch 3/10
157/157 [==============================] - 39s 249ms/step - loss: 0.2686 - accuracy: 0.8964 - val_loss: 0.3208 - val_accuracy: 0.8702
Epoch 4/10
157/157 [==============================] - 41s 260ms/step - loss: 0.2369 - accuracy: 0.9093 - val_loss: 0.3743 - val_accuracy: 0.8486
Epoch 5/10
157/157 [==============================] - 41s 261ms/step - loss: 0.2180 - accuracy: 0.9182 - val_loss: 0.3713 - val_accuracy: 0.8416
Epoch 6/10
157/157 [==============================] - 42s 269ms/step - loss: 0.2043 - accuracy: 0.9232 - val_loss: 0.3295 - val_accuracy: 0.8732
Epoch 7/10
157/157 [==============================] - 42s 267ms/step - loss: 0.1883 - accuracy: 0.9302 - val_loss: 0.3110 - val_ac

In [17]:
from keras.layers import GRU

model_deep = Sequential()
model_deep.add(Embedding(max_features, 32))
model_deep.add(GRU(32, return_sequences=True))  
model_deep.add(GRU(32)) 
model_deep.add(Dense(1, activation='sigmoid'))

In [18]:
model_deep.compile(optimizer='rmsprop', loss='binary_crossentropy', metrics=['accuracy'])
model_deep.fit(x_train, y_train, epochs=10, batch_size=128, validation_split=0.2)

Epoch 1/10
157/157 [==============================] - 54s 317ms/step - loss: 0.5157 - accuracy: 0.7335 - val_loss: 0.3528 - val_accuracy: 0.8462
Epoch 2/10
157/157 [==============================] - 50s 318ms/step - loss: 0.3285 - accuracy: 0.8616 - val_loss: 0.6458 - val_accuracy: 0.7076
Epoch 3/10
157/157 [==============================] - 57s 363ms/step - loss: 0.2835 - accuracy: 0.8863 - val_loss: 0.4019 - val_accuracy: 0.8572
Epoch 4/10
157/157 [==============================] - 57s 365ms/step - loss: 0.2510 - accuracy: 0.9003 - val_loss: 0.3357 - val_accuracy: 0.8562
Epoch 5/10
157/157 [==============================] - 54s 346ms/step - loss: 0.2289 - accuracy: 0.9100 - val_loss: 0.2885 - val_accuracy: 0.8844
Epoch 6/10
157/157 [==============================] - 57s 363ms/step - loss: 0.2176 - accuracy: 0.9162 - val_loss: 0.3505 - val_accuracy: 0.8644
Epoch 7/10
157/157 [==============================] - 54s 342ms/step - loss: 0.1998 - accuracy: 0.9246 - val_loss: 0.3072 - val_ac

In [19]:
loss, accuracy = model_rnn.evaluate(x_test, y_test)
print("RNN Accuracy:", accuracy)

loss, accuracy = model_lstm.evaluate(x_test, y_test)
print("LSTM Accuracy:", accuracy)

loss, accuracy = model_bidirectional.evaluate(x_test, y_test)
print("Bidirectional Accuracy:", accuracy)

loss, accuracy = model_deep.evaluate(x_test, y_test)
print("Deep RNN Accuracy:", accuracy)

782/782 [==============================] - 12s 15ms/step - loss: 0.5739 - accuracy: 0.8072
RNN Accuracy: 0.8071600198745728
782/782 [==============================] - 19s 24ms/step - loss: 0.3848 - accuracy: 0.8671
LSTM Accuracy: 0.8671200275421143
782/782 [==============================] - 30s 39ms/step - loss: 0.3347 - accuracy: 0.8732
Bidirectional Accuracy: 0.8731600046157837
782/782 [==============================] - 44s 56ms/step - loss: 0.3356 - accuracy: 0.8750
Deep RNN Accuracy: 0.8750399947166443


Заголом якщо підібрати параметри і надати більше епох для навчання, точність кожної мережі можна підвищити, але за однакових умов маємо наступні результати:

Проста рекурентна нейронна мережа (RNN) показала точність на рівні 80.72%. Вона є найпростішою моделлю серед усіх розглянутих, але вона може мати проблеми з урахуванням довгострокових залежностей у тексті. 

Модель з використанням LSTM шарів показала кращу точність, досягнувши 86.71%. LSTM має додаткові механізми для збереження та використання довгострокової пам'яті, що робить його ефективним у вирішенні завдань, пов'язаних з послідовностями. 

Bidirectional показала ще кращі результати з точністю 87.32%. Ця модель використовує інформацію з обох напрямків послідовності, що дозволяє краще ураховувати контекст і отримувати більш точні прогнози. 

Deep RNN також показала високу точність 87.50%. Використання глибини у моделі може допомогти краще узагальнити і засвоїти складні залежності у тексті. 

Загалом, LSTM, двостороння рекурентна нейронна мережа та глибока рекурентна нейронна мережа продемонстрували кращу точність у порівнянні з простою RNN.